src/main/scala/is/hail/utils/FlipbookIterator.scala

```diff
@@ -162,22 +162,27 @@ abstract class FlipbookIterator[A] extends BufferedIterator[A] { self =>
 
   def staircased(ord: OrderingView[A]): StagingIterator[FlipbookIterator[A]] = {
     ord.setBottom()
-    val stepIterator: FlipbookIterator[A] = FlipbookIterator(
-      new StateMachine[A] {
-        def value: A = self.value
-        def isValid: Boolean = self.isValid && ord.isEquivalent(value)
-        def advance() = { self.advance() }
+    val stepSM = new StateMachine[A] {
```

Comment:
@patrick-schultz FYI, we did this because the staircase would finish, be pointing at the first element of the new staircase, then we'd clear the region, then we'd try to exhaust when we advance the outer iterator and things would fail. The trick is: we already know if we've reached the end of the staircase, we just need to cache that knowledge before the region is cleared.